### PR TITLE
Specify noDerivative and zeroDerivative.

### LINKED ISSUE #2840: Mathematical formulae for noDerivative and zeroDerivative
Would it be possible to include mathematical formulae describing the meaning of `noDerivative` and `zeroDerivative` annotations for the derivative? The chain rule for a multivariate function is:
image

If we specifying in `f` annotation `derivative(noDerivative = x) = f_der1` what would be the formula for `f_der1`? What would it be if you say `derivative(zeroDerivative = x) = f_der2`?

### PR DESCRIPTION
**Pull request by HansOlsson**
Closes #2840
(And using those formulas, but reworked based on the examples.)

### CODE REVIEW

File: chapters/functions.tex [hunk 1]
```diff
@@ -1459,6 +1465,14 @@ \subsection{Using the Derivative Annotation}\label{using-the-derivative-annotati
 \end{lstlisting}
 This is useful if \lstinline!g! represents the major computational
 effort of \lstinline!fg!.
+
+Therefore \lstinline!f_der! indirectly includes the derivative with respect to \lstinline!y! as follows:
+\begin{eqnarray*}
+\frac{d}{dt}fg(x(t))&=&\frac{d}{dt}f(x(t),g(x(t)))\\
+&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{\partial g}{\partial x}\frac{dx}{dt}\\
+&=&(\frac{\partial f}{\partial x}+\frac{\partial f}{\partial y}\frac{\partial g}{\partial x})\frac{dx}{dt}\\
```
Comment (eshmoylova) [inline chapters/functions.tex:1473]: Might be better to use adjustable parentheses:
```suggestion
&=&\left(\frac{\partial f}{\partial x}+\frac{\partial f}{\partial y}\frac{\partial g}{\partial x}\right)\frac{dx}{dt}\\
```

**Review by eshmoylova — CHANGES_REQUESTED** [review]
I don't know if it's an issue with the example, with the formula, or with the definition of `noDerivative`.  So for now I am just adding a comment here. May need to open a separate issue. The issue is with the `noDerivative` annotation. Looking at the example and the formula in the PR, `f_der` depends on `g`. The annotation "annotation(derivative(noDerivative = y) = f_der);" is specified on `f` without any information about `g`. How can one know which `g` `f_der` is valid for by just looking at `f` and the annotation?

File: chapters/functions.tex [hunk 2]
```diff
@@ -1414,6 +1414,12 @@ \subsection{Using the Derivative Annotation}\label{using-the-derivative-annotati
   //annotation(derivative(order=2) = f_general_der2);
 end f_general_der;
 \end{lstlisting}
+In the example above \lstinline!zeroDerivative=y! imply that
+\begin{eqnarray*}
+\frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}\\
```
[the remark this answers is not in the document]
Comment (HansOlsson) [reply chapters/functions.tex:1419]: > I don't know if it's an issue with the example, with the formula, or with the definition of `noDerivative`. So for now I am just adding a comment here. May need to open a separate issue. The issue is with the `noDerivative` annotation. Looking at the example and the formula in the PR, `f_der` depends on `g`. The annotation "annotation(derivative(noDerivative = y) = f_der);" is specified on `f` without any information about `g`. How can one know which `g` `f_der` is valid for by just looking at `f` and the annotation?

You are right: It cannot.
We did at some point attempt to specify it by some more elaborate variant of noDerivative - but it was too complicated.

File: chapters/functions.tex [hunk 3]
```diff
@@ -1414,6 +1414,12 @@ \subsection{Using the Derivative Annotation}\label{using-the-derivative-annotati
   //annotation(derivative(order=2) = f_general_der2);
 end f_general_der;
 \end{lstlisting}
+In the example above \lstinline!zeroDerivative=y! imply that
```
[the remark this answers is not in the document]
Comment (HansOlsson) [reply chapters/functions.tex]: Of course

File: chapters/functions.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (HansOlsson) [reply chapters/functions.tex]: > Should we include `offset` into the formula as well for consistency? Or remove it from the Modelica code for `f` for consistency and brevity?

I added it (but renamed to "o"). The reason to keep it is that it shows zeroDerivative for multiple variables.

File: chapters/functions.tex [hunk 4]
```diff
@@ -1459,6 +1465,14 @@ \subsection{Using the Derivative Annotation}\label{using-the-derivative-annotati
 \end{lstlisting}
 This is useful if \lstinline!g! represents the major computational
 effort of \lstinline!fg!.
+
+Therefore \lstinline!f_der! indirectly includes the derivative with respect to \lstinline!y! as follows:
+\begin{eqnarray*}
+\frac{d}{dt}fg(x(t))&=&\frac{d}{dt}f(x(t),g(x(t)))\\
+&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{\partial g}{\partial x}\frac{dx}{dt}\\
+&=&(\frac{\partial f}{\partial x}+\frac{\partial f}{\partial y}\frac{\partial g}{\partial x})\frac{dx}{dt}\\
+&=&\textrm{f\_der}(x,y)\frac{dx}{dt}
```
[the remark this answers is not in the document]
Comment (HansOlsson) [reply chapters/functions.tex]: Since `f_der` is not merely the derivative of `f` I renamed it to `h`; and that also corrected the font-issue.

File: chapters/functions.tex [hunk 5]
```diff
@@ -1414,6 +1414,12 @@ \subsection{Using the Derivative Annotation}\label{using-the-derivative-annotati
   //annotation(derivative(order=2) = f_general_der2);
 end f_general_der;
 \end{lstlisting}
+In the example above \lstinline!zeroDerivative=y! implies that
+\begin{eqnarray*}
+\frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}+\frac{\partial f}{\partial o}\frac{do}{dt}\\
```
Comment (eshmoylova) [inline chapters/functions.tex:1419]: `o` is missing:
```suggestion
\frac{d}{dt}f(x(t),y(t),o(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}+\frac{\partial f}{\partial o}\frac{do}{dt}\\
```

File: chapters/functions.tex [hunk 6]
```diff
@@ -1414,6 +1414,12 @@ \subsection{Using the Derivative Annotation}\label{using-the-derivative-annotati
   //annotation(derivative(order=2) = f_general_der2);
 end f_general_der;
 \end{lstlisting}
+In the example above \lstinline!zeroDerivative=y! implies that
```
Comment (eshmoylova) [inline chapters/functions.tex:1417]: `offset` is missing:
```suggestion
In the example above \lstinline!zeroDerivative=y! and \lstinline!zeroDerivative=offset! imply that
```

File: chapters/functions.tex [hunk 7]
```diff
@@ -1414,6 +1414,12 @@ \subsection{Using the Derivative Annotation}\label{using-the-derivative-annotati
...
+\begin{eqnarray*}
+\frac{d}{dt}f(x(t),y(t))&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\frac{dy}{dt}+\frac{\partial f}{\partial o}\frac{do}{dt}\\
+&=&\frac{\partial f}{\partial x}\frac{dx}{dt}+\frac{\partial f}{\partial y}\cdot 0+\frac{\partial f}{\partial o}\cdot 0\\
+&=&\frac{\partial f}{\partial x}\frac{dx}{dt}
```
Comment (eshmoylova) [inline chapters/functions.tex:1421]: Should we include `f_der` into the formula? 
```suggestion
&=&\frac{\partial f}{\partial x}\frac{dx}{dt}\\
&=&f\_der
```
Or it can be on the same line?

**Review by eshmoylova — CHANGES_REQUESTED** [review]

File: chapters/functions.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (eshmoylova) [reply chapters/functions.tex]: > > I don't know if it's an issue with the example, with the formula, or with the definition of `noDerivative`. So for now I am just adding a comment here. May need to open a separate issue. The issue is with the `noDerivative` annotation. Looking at the example and the formula in the PR, `f_der` depends on `g`. The annotation "annotation(derivative(noDerivative = y) = f_der);" is specified on `f` without any information about `g`. How can one know which `g` `f_der` is valid for by just looking at `f` and the annotation?
> 
> You are right: It cannot.
> We did at some point attempt to specify it by some more elaborate variant of noDerivative - but it was too complicated.

That makes it a rather useless annotation then, in my opinion. If a tool does not know in which cases it should be used, the safest way is not to use it.

[the remark this answers is not in the document]
Comment (HansOlsson) [reply chapters/functions.tex]: On one hand it's true that it's unsafe and best to not use - on the other hand it's quite a common case for the functions Modelica.Media due to their construction. At some point we should investigate if that could be improved.

File: chapters/functions.tex (hunk 7 above)
Comment (HansOlsson) [reply chapters/functions.tex:1421]: I would prefer not to include `f_der` at this time, as the underscore looks misplaced in the pdf-file. 
The next option would be to rename f_der in this case as well (e.g. to h) - which should work although a bit odd.

File: chapters/functions.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (eshmoylova) [reply chapters/functions.tex]: > On one hand it's true that it's unsafe and best to not use - on the other hand it's quite a common case for the functions Modelica.Media due to their construction. At some point we should investigate if that could be improved.

OK, I just want to double check that the formula and the example are correct and as intended. In the original response to my ticket #2840 there was no mention of `fg`. So I understood it that we always assume that `y` is a function of `g`. From this PR I see that `y` can be independent of `x` and `h` depends on what `g` is. So I just want to confirm that the way it is in the PR is the way it is meant to be.

I see that in Modelica.Media `Water.IF97_Utilities.T_props_ph`, for example, has a `derivative` annotation with `noDerivative` specified on the third argument, and that `T_props_ph` is always called  as `T_props_ph(p,h,waterBaseProp_ph(p,h,phase,region)`. So, I guess, the assumption is that whenever we need a derivative and it is called with the noDerivative argument as a function of other arguments we can use the derivative in the annotation?

File: chapters/functions.tex [hunk 8]
```diff
@@ -1360,9 +1360,9 @@ \subsection{Using the Derivative Annotation}\label{using-the-derivative-annotati
 Assume that function \lstinline!f! takes a matrix and a scalar.
 Since the matrix argument is usually a parameter expression it is then
 useful to define the function as follows (the additional derivative =
-\lstinline!f_general_der! is optional and can be used when the derivative of
-the matrix or offset is non-zero). Note that \lstinline!f_der! must have
-\lstinline!zeroDerivative! for both \lstinline!y! and \lstinline!offset!, but \lstinline!f_general_der! shall not have
+\lstinline!fGeneralDer! is optional and can be used when the derivative of
+the matrix or offset is non-zero). Note that \lstinline!fDer! must have
+\lstinline!zeroDerivative! for both \lstinline!y! and \lstinline!offset!, but \lstinline!fGeneralDer! shall not have
 \lstinline!zeroDerivative! for either of them (it may \lstinline!zeroDerivative! for \lstinline!x_der!,
```
Comment (eshmoylova) [inline chapters/functions.tex:1366]: It looks like there is a verb missing in "(it may \lstinline!zeroDerivative! for \lstinline!x_der!" Same problem as the comment above: the sentence was there before this PR. Should I open a separate issue?

File: chapters/functions.tex [hunk 9]
```diff
@@ -1360,9 +1360,9 @@ \subsection{Using the Derivative Annotation}\label{using-the-derivative-annotati
 Assume that function \lstinline!f! takes a matrix and a scalar.
 Since the matrix argument is usually a parameter expression it is then
 useful to define the function as follows (the additional derivative =
-\lstinline!f_general_der! is optional and can be used when the derivative of
-the matrix or offset is non-zero). Note that \lstinline!f_der! must have
-\lstinline!zeroDerivative! for both \lstinline!y! and \lstinline!offset!, but \lstinline!f_general_der! shall not have
+\lstinline!fGeneralDer! is optional and can be used when the derivative of
+the matrix or offset is non-zero). Note that \lstinline!fDer! must have
+\lstinline!zeroDerivative! for both \lstinline!y! and \lstinline!offset!, but \lstinline!fGeneralDer! shall not have
```
Comment (eshmoylova) [inline chapters/functions.tex:1365]: What does it mean: "must have zeroDerivative for both y and offset"? In the annotation for the derivative of fDer? Sorry, I realize that it is not really part of the PR because the text was there before. Just reviewing PRs brings attention to some things that I would otherwise disregard when simply reading. Should I open a separate issue?

File: chapters/functions.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (HansOlsson) [reply chapters/functions.tex]: Yes.

[the remark this answers is not in the document]
Comment (eshmoylova) [reply chapters/functions.tex]: OK. Resolving the conversation.

**Review by henrikt-ma — COMMENTED** [review]
With @eshmoylova doing a good job commenting on the content, I thought I should try to at least take a look at the form. Unfortunately, I find it almost impossible to do this from where I am now, using just a phone. I think I can see some minor math that can be improved, but I'll have to postpone leaving any comments with real detail until after this week.

File: chapters/functions.tex (hunk 7 above)
Comment (HansOlsson) [reply chapters/functions.tex]: I renamed to fDer, and included it.

**Review by eshmoylova — APPROVED** [review]
Looks good to me.
